reproman/resource/tests/test_singularity.py

```diff
@@ -10,77 +10,92 @@
+    with pytest.raises(TypeError):
+        Singularity(name='foo')
 
-    # Set working directory to a scratch directory since we will be creating
-    # Singularity image files during testing.
-    orig_cwd = os.getcwd()
-    tempdir = tempfile.mkdtemp()
-    os.chdir(tempdir)
 
+@mark.skipif_no_network
+@mark.skipif_no_singularity
+def test_singularity_resource_class(tmpdir):
+    tmpdir = str(tmpdir)
     with swallow_logs(new_level=logging.DEBUG) as log:
+        Runner(cwd=tmpdir).run(
+            ['singularity', 'pull', '--name', 'img',
+             'shub://truatpasteurdotfr/singularity-alpine'])
 
+        # ATTN: Apparently an instance name can't contain a hyphen.
+        name = "reproman_test_{}".format(str(uuid.uuid4())[:4])
+        image = op.join(tmpdir, 'img')
         # Test creating a new singularity container instance.
-        resource = Singularity(name='foo',
-            image='shub://truatpasteurdotfr/singularity-alpine')
-        assert resource.name == 'foo'
-        assert resource.image == 'shub://truatpasteurdotfr/singularity-alpine'
+        resource = Singularity(name=name, image=image)
+        assert resource.name == name
+        assert resource.image == image
         resource.connect()
         assert resource.id is None
         assert resource.status is None
         list(resource.create())
-        assert resource.id.startswith('foo-')
-        assert resource.status == 'running'
-
-        # Test trying to create an already running instance.
-        resource_duplicate = Singularity(name='foo',
-            image='shub://truatpasteurdotfr/singularity-alpine')
-        resource_duplicate.connect()
-        assert resource_duplicate.id.startswith('foo-')
-        assert resource_duplicate.status == 'running'
-        list(resource_duplicate.create())
-        assert_in("Resource 'foo' already exists.", log.lines)
-
-        # Test retrieving instance info.
-        info = resource.get_instance_info()
-        assert info['name'] == 'foo'
-        assert re.match(r'^\d+$', info['pid'])
-        assert info['image'].endswith('.simg')
-
-        # Test starting an instance.
-        with pytest.raises(NotImplementedError):
-            resource.start()
-
-        # Test stopping an instance.
-        with pytest.raises(NotImplementedError):
-            resource.stop()
+        to_delete = [resource]
+        try:
+            assert resource.id.startswith(name + "-")
+            assert resource.status == 'running'
+
+            # Test trying to create an already running instance.
+            resource_duplicate = Singularity(name=name, image=image)
+            resource_duplicate.connect()
+            assert resource_duplicate.id.startswith(name + "-")
+            assert resource_duplicate.status == 'running'
+            list(resource_duplicate.create())
+            assert any("Resource '{}' already exists".format(name) in l
+                       for l in log.lines)
```

Comment:
Why this way rather than using tests.utils.assert_in() ?

Reply:
> Why this way rather than using tests.utils.assert_in() ?

Short answer: assert_in doesn't offer any advantage over `assert ... in ...`, but you're right that the form I chose isn't good.  Will update.

Longer: With d3b16d4b5 (TST: Add pytest replacements for nose.tools helpers, 2018-06-25), these helpers were kept around for compatibility and to avoid code churn.  As the code comment from that commit says, I think we should generally avoid these in new code because they don't take advantage of pytest's assertion rewriting.  We could tell pytest to rewrite assertions in the test.utils module with pytest.register_assert_rewrite, but I didn't bother since most of these helpers are replaced adequately by plain assertions + pytest's rewriting.

That being said, you're right that what I did here is **not** an improvement because the message by pytest doesn't provide any details about the failure.  If I had done the faithful conversion of

```
assert "Resource '{}' already exists.".format(name) in log.lines
```

the error reporting would have been as good as assert_in.  I chose to do the `any` bit and match within strings of log.lines because I thought matching on a period was unnecessarily brittle.

So I could (1) use the snippet above with the period, (2) drop the period and use assert_re_in, or (3) drop the period and use `assert ... in log.out`.  I'll go with 3.